appsrc/acrnema/ancreate.man
.TH ANCREATE 1 "19 Jan 2007" "DICOM PS3" "DICOM PS3 - Create dicom file"
.SH NAME
ancreate \- ACR/NEMA DICOM PS3 ... Create dicom file
.SH SYNOPSIS
.HP 10
.B ancreate
[
.B \-b|big
]
[
.B \-e|explicit
]
[
.B \-p|preamble
]
[
.B \-v|verbose
]
<
.I inputfilename
>
.I outputfilename
.SH DESCRIPTION
.LP
.B ancreate
creates a dicom 3 or earlier acr/nema file from a description of attributes. The description of attributes matches the output of dcdump, so that pieces of dumped files can be edited and files reconstructed. Note that the library of dicom attribute handling, dictionary, and transfer syntax routines are NOT used in this program, which allows it to be used as a standalone program to test the library and other utilities.
.LP
Everything needed to create a file is specified in a description read from standard input, apart from some very simple rules that recognize certain value representations for special treatment of the output format (eg. OW, OB and SQ in explicit value representation transfer syntaxes are handled differently).
.LP
The description consists of lines of the following format:
.LP
.B attributes  ...
.RE

.RE
(group,element) .* VR=<vr> VL=<vl> [vals]
.RE
(group,element) .* VR=<vr> VL=<vl> <vals>
.RE
(group,element) .* VR=<FD|FL> VL=<vl> {vals}
.RE

Note that the group, element, value length, and values are always in hexadecimal format, whether a leading "0x" is present or not. Hence "14" and "014" are both interpreted as "0x14", or 20 decimal, in these positions (unlike data descriptions). Multiple values should be delimited with a '\\' for string values (CS, DS. etc.) and a ',' for binary attributes (US, etc.).
Make sure there are no spaces between values or the value length will be computed incorrectly (i.e. the number of values are computed from the number of non-value characters).
.LP
The value length is measured against the value length specified, and a warning is issued if they are different. The actual value, rather than the specified value is used, for all except value representations of OW, OB and SQ. This allows the special value of 0xffffffff to be passed (which means "unspecified" to dicom), as well as allowing the actual values to be included on separate lines following the attribute description. Remember that the value length is in bytes, not the number of values, and must be correct if not 0xffffffff, and even the latter is only supported in DICOM for certain encapsulated transfer syntaxes such as JPEG, and not the normal and default transfer syntaxes.
.LP
If the values for an OB or OW attribute are to be specified on suceeding lines, remember to include an empty value list "[]" or the parser will freak.
.LP
An empty value list like "[]" is also required after each SQ (Sequence) element, even though the "payload" of the
sequence follows enclosed in delimited items.
.LP
Anything after a # either at the start of a line, or after the value of a data element, will be treated as
a comment and skipped.
.LP
White space may preceed the the first symbol of a command or data element.
.LP
.B transfer syntax commands ...
.RE

.RE
%implicit
.RE
%explicit
.RE
%little
.RE
%big
.RE

Note that the use of big always forces explicit value representation, as there is no valid "implicit value representation big endian" transfer syntax in dicom. These commands are present here, as well as in the command-line options, to allow a change in transfer syntax, such as after a metaheader.
.LP
.B sequence commands ...
.RE

.RE
%item
.RE
%enditem
.RE
%endseq
.RE

Use these commands to insert contents of undefined length sequences - remember to set
the VL of the enclosing sequence attribute to 0xFFFFFFFF and include an empty value "[]".
Note that the "-ancreate" option of the andump utility can be used to generate these.

.LP
.B data value format commands ...
.RE

.RE
%byte
.RE
%short
.RE
%word
.RE
%long
.RE

Specify the interpretation of subsequent data values, as 8, 16 or 32 bit values. Short and word are synonymous. Use of these commands is usually unnecessary, as the presence of an OB or OW value representation in an attribute description automagically sets the data type to byte or short as appropriate. Prior to such an attribute, and in the absence of a data value format command, the default type is byte.
.LP
.B data ...
.RE

.RE
hexvalue(s)
.RE
octalvalue(s)
.RE
decvalue(s)
\n
Note that the C convention of hexvalues starting with "0x", octal values starting with "0", and others being decimal values is followed for data values (unlike attribute descriptions). The data values are included in the output at the current position.
.LP
.SH OPTIONS
Only redirection from standard input or to standard output are
supported. Input and output file names may not
be specified. Errors and warnings goes to standard error.
.LP
The default transfer syntax is implicit value representation, little endian.
.PP
Options specific to this program are:
.TP
.B \-b|big
.RS
Big endian transfer syntax.
.RE
.TP
.B \-e|explicit
.RS
Explicit value representation transfer syntax.
.RE
.TP
.B \-p|preamble
.RS
Include file preamble (128 bytes of zero plus "DICM" magic string) before writing anything.
Note that this option is completely independent of whether or not any meta information header
(group 0x0002) elements are manually added from the input stream or not.
.RE
.TP
.B \-v|verbose
.RS
Dump data element tags and command numnber (not precisely the same as
the input line number) to stderr while parsing. Helps to localize errors
in the input.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
Sample attributes that can be fed to ancreate:
.RE

.RE
% ancreate >test
.RE
(0000x8,0000x8) VR=<CS> VL=<0x14> <ORIGINAL\\PRIMARY\\MPR> 
.RE
(0000x8,000x16) VR=<UI> VL=<0x10> <1.2.840.10008.0> 
.RE
(0x0018,0x9104) VR=<FD> VL=<0x0010> {13.69,1.36e-08} 
.RE
(000x28,0000x0) VR=<UL> VL=<0x4> [0xb6]
.RE
(000028,000002) VR=<US> VL=<2> [1]
.RE
(000028,000004) VR=<CS> VL=<c> <MONOCHROME2> 
.RE
(0x6000,000x50) VR=<SS> VL=<0x4> [0x5c31,0x2031]
.LP
.RE
Sample use of unspecified vl ...
.RE

.RE
% ancreate >test
.RE
(0x7fe0,000x10) OX PixelData VR=<OW> VL=<0xffffffff> []
.LP
.RE
Sample use of data values ...
.RE

.RE
% ancreate >test
.RE
(0x7fe0,000x10) OX PixelData VR=<OW> VL=<0x8> []
.RE
0x10 0x12
.RE
0x24 0x36
.RE

.RE
% dcdump test
.RE
(0x7fe0,000x10) OX PixelData     VR=<OW>   VL=<0x8> 
.RE

.RE
% od -x test
.RE
0000000  e07f 1000 0800 0000 1000 1200 2400 3600
.RE
0000020
.LP
.RE

.RE
Sample use of metaheader, then explicit big syntax ...
.RE

.RE
% ancreate >test
.RE
(0000x2,000x10) UI TransferSyntaxUID VR=<UI> VL=<0x14>
.RE
                <1.2.840.10008.1.2.2>
.RE
%big
.RE
(000028,000002) VR=<US> VL=<2> [1]
.LP
.RE
Should be able to find no differences:
.RE

.RE
% dcdump test 2>test.dump
.RE
% ancreate <test.dump >test2
.RE
% cmp test test2
.RE
% dcdump test2 2>test2.dump
.RE
% diff -b test.dump test2.dump
.RE

.RE
(unless of course there are OB or OW attributes, in which case the data will be elided, and the comparison will fail, or there is a metaheader present which changes the transfer syntax in the middle of the file).
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR andump(1) ,
.BR dcdump(1) ,
.BR dcintro(1) ,
.BR dcintro(3) ,
.BR dcintro(5)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
In the attribute descriptions, the group, element, value length, and values are always in hexadecimal format, whether a leading "0x" is present or not. This is because dcdump adopts such a convention, with slightly different output on different platforms and compilers. In particular lib-g++ does not output the "x" in hexadecimal formats :(
.LP
It would be nice to have the verbose output reference actual line numbers rather
than just command numbers.
.LP
It would be nice to have the error messages reference actual line numbers.
.LP
The maximum input line length is 32768 and will be truncated after that causing errors and corruption - this really should be unlimited.



CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/acrnema/ancreate.cc
#include <iostream.h>
#include <iomanip.h>
#include <string.h>
#include <ctype.h>

#include "basetype.h"
#include "txstream.h"
#include "getoptns.h"
#include "mesgtext.h"

#define INPUTLINEBUFFERLENGTH 32768

// all the following is copied from dconvert/fltype.h

#include <float.h>

#ifdef POWINTEGEREXPONENTTYPE
#define	powi(m,e)	pow(m,(POWINTEGEREXPONENTTYPE)(e))
#else
#define	powi(m,e)	pow(m,(long)(e))
#endif

#ifdef __SC__
#define infinity()    HUGE_VAL
#define quiet_nan(x)  0
#endif /* __SC__ */

#ifdef __MWERKS__
#define infinity()    __inf()
#define quiet_nan(x)  0
#endif /* __MWERKS__ */

#ifdef USEDUMBINFINITYANDNAN
#if USEDUMBINFINITYANDNAN
#define infinity()    FLT_MAX
#define quiet_nan(x)  0
#endif
#endif


static void
write32(ostream& out,Uint32 u,int big)
{
//cerr << "write32 0x" << hex << u << dec << endl;
	unsigned char buffer[4];
	if (big) {
		buffer[0]=(unsigned char)(u>>24);
		buffer[1]=(unsigned char)(u>>16);
		buffer[2]=(unsigned char)(u>>8);
		buffer[3]=(unsigned char)u;
	}
	else {
		buffer[3]=(unsigned char)(u>>24);
		buffer[2]=(unsigned char)(u>>16);
		buffer[1]=(unsigned char)(u>>8);
		buffer[0]=(unsigned char)u;
	}
	out.write((char *)buffer,4);
}

static void
write16(ostream& out,Uint16 u,int big)
{
	unsigned char buffer[4];
	if (big) {
		buffer[0]=(unsigned char)(u>>8);
		buffer[1]=(unsigned char)u;
	}
	else {
		buffer[1]=(unsigned char)(u>>8);
		buffer[0]=(unsigned char)u;
	}
	out.write((char *)buffer,2);
}

static void
write8(ostream& out,unsigned char u)
{
	unsigned char buffer[1];
	buffer[0]=u;
	out.write((char *)buffer,1);
}

static void
writeFloat64(ostream& out,double value,int big) {
//cerr << "writeFloat64: " << value << endl;

	// copied from libsrc/include/dctool/attrtypf.h newSize8Value()

	Uint32 sign=0;
	Int32 exponent;
	Uint32 mantissahigh;
	Uint32 mantissalow;

	if (value < 0) {
		sign=0x80000000;
		value=-value;
	}

	if (value == 0) {
		exponent=0;
		mantissahigh=0;
		mantissalow=0;
	}
	else {
		exponent=(int)ourlog2(value);
		if (value < 1) --exponent;
		if (exponent <= -1022) {
			double mantissavalue = value/pow(2,-1022);
			mantissahigh=(Uint32)(mantissavalue*powi(2.0,20));
			mantissalow =(Uint32)(mantissavalue*powi(2.0,52));
			exponent=0;
		}
		else {
			if (exponent > 1023) exponent=1023;
			double mantissavalue = value/pow(2,exponent) - 1.0;
			mantissahigh=(Uint32)(mantissavalue*powi(2.0,20));
			mantissalow =(Uint32)(mantissavalue*powi(2.0,52));
			exponent=((exponent+1023)<<20)&0x7ff00000;
		}
	}

	Uint32 binaryhigh=sign+exponent+mantissahigh;

	unsigned char *barray = new unsigned char[8];
	if (big) {
		barray[0]=(unsigned char)(binaryhigh>>24);
		barray[1]=(unsigned char)(binaryhigh>>16);
		barray[2]=(unsigned char)(binaryhigh>>8);
		barray[3]=(unsigned char)(binaryhigh);
		barray[4]=(unsigned char)(mantissalow>>24);
		barray[5]=(unsigned char)(mantissalow>>16);
		barray[6]=(unsigned char)(mantissalow>>8);
		barray[7]=(unsigned char)(mantissalow);
	}
	else {
		barray[7]=(unsigned char)(binaryhigh>>24);
		barray[6]=(unsigned char)(binaryhigh>>16);
		barray[5]=(unsigned char)(binaryhigh>>8);
		barray[4]=(unsigned char)(binaryhigh);
		barray[3]=(unsigned char)(mantissalow>>24);
		barray[2]=(unsigned char)(mantissalow>>16);
		barray[1]=(unsigned char)(mantissalow>>8);
		barray[0]=(unsigned char)(mantissalow);
	}
	out.write((char *)barray,8);
}

static void
writeFloat32(ostream& out,float value,int big) {
//cerr << "writeFloat32: " << f << endl;

	// copied from libsrc/include/dctool/attrtypf.h newSize4Value()
	
	Uint32 sign=0;
	Int32 exponent;
	Uint32 mantissa;

	if (value < 0) {
		sign=0x80000000;
		value=-value;
	}

	if (value == 0) {
		exponent=0;
		mantissa=0;
	}
	else {
		exponent=(int)ourlog2(value);
		if (value < 1) --exponent;
		if (exponent <= -126) {
			mantissa=(Uint32)((value/pow(2,-126))*(1<<23));
			exponent=0;
		}
		else {
			if (exponent > 127) exponent=127;
			mantissa=(Uint32)(((value/pow(2,exponent))-1)*(1<<23));
			exponent=((exponent+127)<<23)&0x7f800000;
		}
	}

	Uint32 binary=sign+exponent+mantissa;

	unsigned char *barray = new unsigned char[4];
	if (big) {
		barray[0]=(unsigned char)(binary>>24);
		barray[1]=(unsigned char)(binary>>16);
		barray[2]=(unsigned char)(binary>>8);
		barray[3]=(unsigned char)(binary);
	}
	else {
		barray[3]=(unsigned char)(binary>>24);
		barray[2]=(unsigned char)(binary>>16);
		barray[1]=(unsigned char)(binary>>8);
		barray[0]=(unsigned char)(binary);
	}
	out.write((char *)barray,4);
}

/* Recognized commands are:

(hex group,hex element) anything VR=<vr> VL=<hex vl> [hexvals]
(hex group,hex element) anything VR=<vr> VL=<hex vl> <strvals>
%implicit
%explicit
%little
%big

%item
%enditem
%endseq

datavalues

VL will be replaced with actual VL unless OB,OW or SQ

*/

static void
parseError(const char *str,char c)
{
	cerr << "Parse error in entry - \"" << str 
	     << "\" expected, got \'" << c << "\'\n" << flush;
}

static int
getCommand(istream& in)
{
	int command=0;
	const unsigned buflng = 10;
	char buffer[buflng];
		char c;
		do {
			in >> c;
			if (!in.good()) return 0;
		} while (isspace(c));
		if (c == '#') {					// start of comment
			do {					// skip to end of line
				in >> c;
				if (!in.good()) return 0;
			} while (c != '\n' && c != '\r');
			do {					// skip ends of lines
				in >> c;
				if (!in.good()) return 0;
			} while (c == '\n' || c == '\r');
		}
		// c should now be non-white space beginning of command
		switch (c) {
			case '(':
				command='(';
				break;
			case '%':
				in >> setw(buflng-1) >> buffer;
				if (!in.good()) break;
				if (strcmp(buffer,"implicit") == 0) command='i';
				if (strcmp(buffer,"explicit") == 0) command='e';
				if (strcmp(buffer,"little") == 0)   command='l';
				if (strcmp(buffer,"big") == 0)      command='b';
				if (strcmp(buffer,"byte") == 0)     command='B';
				if (strcmp(buffer,"short") == 0)    command='S';
				if (strcmp(buffer,"word") == 0)     command='S';
				if (strcmp(buffer,"long") == 0)     command='L';
				if (strcmp(buffer,"item") == 0)     command='T';
				if (strcmp(buffer,"enditem") == 0)  command='M';
				if (strcmp(buffer,"endseq") == 0)   command='Q';
				break;
			default:
				if (isdigit(c)) {
					command=c;
					in.putback(c);
				}
				else {
					parseError("attribute,command or value",
						c);
				}
				break;
		}
	return command;
}

static void
getFloat64FromString(istream& in,double &d) {
	in >> dec >> d;
}

static void
getFloat32FromString(istream& in,float &f) {
	in >> dec >> f;
}

static void
getHex32(istream& in,Uint32& v)
{
	char c;
	int i;

	in >> hex >> v;
	if (v == 0 && (i=in.peek()) == 'x') {
		in.get(c);
		in >> hex >> v;
	}
	// this check seems to fail on trailing ']' delimiter with gcc 3 :(
	//if (!in.good()) {
	//	parseError("hex number",0);
	//}
}

static void
getHex16(istream& in,Uint16& v)
{
	Uint32 v32;
	getHex32(in,v32);
	v=(Uint16)v32;
}


static int
getParsedEntry(istream& in,Uint16& group,Uint16& element,
	char *VR,Uint32& VL,int& valuetype,char **valuestring,bool verbose,int commandcount)
{
	char c;
	const unsigned buflng = INPUTLINEBUFFERLENGTH;
	char buffer[buflng];	// enough for longest value string
	char *p;

	// First '(' has already been read as command

	// Find "nnnn,nnnn)" ...

	in >> resetiosflags(ios::basefield) >> group;
	if (!in.good()) {
		return -1;	// Failure
	}
	in >> c;
	if (c != ',') {
		parseError(",",c);
		return -1;	// Failure
	}
	in >> resetiosflags(ios::basefield) >> element;
	if (!in.good()) {
		return -1;	// Failure
	}
	in >> c;
	if (c != ')') {
		parseError(")",c);
		return -1;	// Failure
	}

	if (verbose) {
		cerr << dec << commandcount << ": (";
		writeZeroPaddedHexNumber(cerr,group,4);
		cerr << ",";
		writeZeroPaddedHexNumber(cerr,element,4);
		cerr << ") " << endl;
	}

	// Skip any fields (eg. from dc3dump) found "VR=<xx>" ...
	do {
		in >> buffer;
	} while (in.good() && strncmp(buffer,"VR",2) != 0);

	if (!in.good()) {
		parseError("VR",0);
		return -1;	// Failure
	}
	for (p=buffer; *p && *p != '<';++p);
	strncpy(VR,p+1,2);
	VR[2]=0;
	if (*(p+3) != '>') {
		parseError("> after VR",*(p+3));
		return -1;	// Failure
	}

	// Find "VL=<nn>" ... 	(nn is in hex)

	do {
		in >> buffer;
	} while (in.good() && strncmp(buffer,"VL",2) != 0);

	if (!in.good()) {
		parseError("VL",0);
		return -1;	// Failure
	}
	for (p=buffer; *p && *p != '<';++p);
	istrstream istr(p+1);
	istr >> resetiosflags(ios::basefield) >> VL;
//cerr << "vl=" << VL << endl;
	if (!istr.good()) {
		return -1;	// Failure
	}
 	istr >> c;
	if (c != '>') {
		parseError("> after VL",c);
		return -1;	// Failure
	}

	in >> c;
	char traildelim;
	switch (c) {
		case '<':
			traildelim='>';
			break;
		case '[':
			traildelim=']';
			break;
		case '{':
			traildelim='}';
			break;
		default:
			parseError("<values> or [values] or {values}",c);
			return -1;	// Failure
	}
	valuetype=c;
	for (p=buffer; p<buffer+buflng; ++p) {
		in.get(c);	// Don't skip whitespace
		if (!in.good()) {
			parseError("> or ] or } after value",0);
			return -1;	// Failure
		}
		else {
			if (c == traildelim) {
				*p=0;
				break;
			}
			else {
				*p=c;
			}
		}
	}
	*valuestring=new char[strlen(buffer)+1];
	strcpy(*valuestring,buffer);

	return 0;		// Success
}

static void
writeAttribute(ostream& out,Uint16 group,Uint16 element,
	char *vr,Uint32 vl,int type,char *string,
	int explicitvr,int big)
{
	write16(out,group,big);
	write16(out,element,big);

	Uint32 actualvl;
	switch (type) {
		case '<':	// string
			if ((actualvl=strlen(string))%2) ++actualvl;
			break;
		case '[':	// hex
			if (*string) {
				char *p;
				for (actualvl=1,p=string; *p; ++p)
					if (!isxdigit(*p) && *p != 'x' && !isspace(*p)) 
						++actualvl;
			}
			else actualvl=0;
			actualvl*=(vr[1] == 'L')?4:2;
			break;
		case '{':	// float
			if (*string) {
				char *p;
				for (actualvl=1,p=string; *p; ++p)
					if (!isdigit(*p) && *p != '.' && *p != '-' && *p != 'e' && *p != 'E' && !isspace(*p)) 
						++actualvl;
			}
			else actualvl=0;
			actualvl*=(vr[1] == 'D')?8:4;
			break;
	}

	if (actualvl != vl) {
		if (strcmp(vr,"OB") == 0
		 || strcmp(vr,"OF") == 0
		 || strcmp(vr,"OW") == 0
		 || strcmp(vr,"SQ") == 0) {
			actualvl=vl;
		}
		else {
			cerr << "Specified VL ";
			writeZeroPaddedHexNumber(cerr,vl,8);
			cerr << " wrong, using ";
			writeZeroPaddedHexNumber(cerr,actualvl,8);
			cerr << " (VR is " << vr << ")\n" << flush;
			//cerr << "VR[0]=" << (unsigned)(vr[0]) << endl;
			//cerr << "VR[1]=" << (unsigned)(vr[1]) << endl;
			//cerr << "VR[2]=" << (unsigned)(vr[2]) << endl;
		}
	}

	if (explicitvr) {
		out.write(vr,2);			// 2 byte explicitvr vr
		if (strcmp(vr,"OB") == 0
		 || strcmp(vr,"OF") == 0
		 || strcmp(vr,"OW") == 0
		 || strcmp(vr,"UN") == 0
		 || strcmp(vr,"UT") == 0
		 || strcmp(vr,"SQ") == 0) {
			write16(out,0,big);		// 2 byte reserved
			write32(out,actualvl,big);	// 4 byte value length
		}
		else {
			write16(out,(Uint16)actualvl,big);	// 2 byte length
		}
	}
	else {
		write32(out,actualvl,big);
	}

	char pad = (strcmp(vr,"UI") == 0) ? 0 : ' ';
	switch (type) {
		case '<':	// string
			out << string;
			if (strlen(string)%2) out << pad;
			break;
		case '[':	// hex
			if (*string) {
				istrstream istr(string);
				while (istr.good()) {
					Uint32 u;
					char c;
					getHex32(istr,u);
					istr >> c;		// delimiter
					if (vr[1] == 'B')
						write8(out,(unsigned char)u);
					else if (vr[1] == 'L')
						write32(out,u,big);
					else
						write16(out,(Uint16)u,big);
				}
			}
			break;
		case '{':	// float
			if (*string) {
				istrstream istr(string);
				while (istr.good()) {
					if (vr[1] == 'D') {
						double d;
						getFloat64FromString(istr,d);
						char c;
						istr >> c;		// delimiter
						writeFloat64(out,d,big);
					}
					else {
						float f;
						getFloat32FromString(istr,f);
						char c;
						istr >> c;		// delimiter
						writeFloat32(out,f,big);
					}
				}
			}
			break;
	}
}

static void
writeItem(ostream& out,int explicitvr,int big)
{
	write16(out,0xfffe,big);
	write16(out,0xe000,big);
	write32(out,0xffffffff,big);
}

static void
writeItemDelimiter(ostream& out,int explicitvr,int big)
{
	write16(out,0xfffe,big);
	write16(out,0xe00d,big);
	write32(out,0x00000000,big);
}

static void
writeSequenceDelimiter(ostream& out,int explicitvr,int big)
{
	write16(out,0xfffe,big);
	write16(out,0xe0dd,big);
	write32(out,0x00000000,big);
}

static void
writePreamble(ostream& out)
{
	char zeroes[128];
	memset(zeroes,0,128);
	out.write(zeroes,128);
	out.write("DICM",4);
}

int
main(int argc, char *argv[])
{
	GetNamedOptions options(argc,argv);
	bool explicitvr	=options.get("explicit")||options.get("e");
	bool big	=options.get("big")||options.get("b");
	if (big) explicitvr=true;

	bool preamble=options.get("preamble")||options.get("p");

	bool verbose=options.get("verbose")||options.get("v");

	options.done();

	cerr << options.errors();

	if (!options.good() || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< " [-b|big]"
			<< " [-e|explicit]"
			<< " [-p|preamble]"
			<< " [-v|verbose]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	if (preamble) {
		writePreamble(cout);
	}

	int datawordlength=1;
	int commandcount=0;
	int command;
	while (command=getCommand(cin)) {
		++commandcount;
		Uint16 group;
		Uint16 element;
		char vr[3];
		Uint32 vl;
		int type;
		char *string;
		switch (command) {
			case '(':
				if (getParsedEntry(cin,group,element,
						vr,vl,type,&string,verbose,commandcount) == 0) {
					writeAttribute(cout,group,element,
						vr,vl,type,string,
						explicitvr,big);
				}
				if (strcmp(vr,"OB") == 0) datawordlength=1;
				if (strcmp(vr,"OW") == 0) datawordlength=2;
				break;
			case 'i':
				explicitvr=0;
				break;
			case 'e':
				explicitvr=1;
				break;
			case 'l':
				big=0;
				break;
			case 'b':
				big=1;
				explicitvr=1;	// no such thing as implicit big
				break;
			case 'B':
				datawordlength=1;
				break;
			case 'S':
				datawordlength=2;
				break;
			case 'L':
				datawordlength=4;
				break;
			case 'T':
				writeItem(cout,explicitvr,big);
				break;
			case 'M':
				writeItemDelimiter(cout,explicitvr,big);
				break;
			case 'Q':
				writeSequenceDelimiter(cout,explicitvr,big);
				break;
			default:
				if (isdigit(command)) {
					Uint32 value;
					cin >> resetiosflags(ios::basefield)
					    >> value;
					switch (datawordlength) {
						case 1:
							cout.put((char)value);
							break;
						case 2:
							write16(cout,
							    (Uint16)value,big);
							break;
						case 4:
							write32(cout,value,big);
							break;
					}
				}
				break;
		}
	}

	return 0;
}

